### PR TITLE
wallet rpc: enable solo mining

### PR DESCRIPTION
**Pull request by stoffu**

### CODE REVIEW

File: src/wallet/wallet_rpc_server.cpp [hunk 1]
```diff
@@ -1424,6 +1427,54 @@ namespace tools
...
+  bool wallet_rpc_server::on_start_mining(const wallet_rpc::COMMAND_RPC_START_MINING::request& req, wallet_rpc::COMMAND_RPC_START_MINING::response& res, epee::json_rpc::error& er)
+  {
+    if (!m_trusted_daemon)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "This command requires a trusted daemon.";
+      return false;
+    }
+
+    size_t max_mining_threads_count = (std::max)(tools::get_max_concurrency(), static_cast<unsigned>(2));
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: Why not allow more than 2 ?

File: src/wallet/wallet_rpc_server.cpp [hunk 2]
```diff
@@ -1424,6 +1427,54 @@ namespace tools
...
+    daemon_req.miner_address = m_wallet.get_account().get_public_address_str(m_wallet.testnet());
+    daemon_req.threads_count        = req.threads_count;
+    daemon_req.do_background_mining = req.do_background_mining;
+    daemon_req.ignore_battery       = req.ignore_battery;
+
+    cryptonote::COMMAND_RPC_START_MINING::response daemon_res;
+    bool r = net_utils::invoke_http_json("/start_mining", daemon_req, daemon_res, m_http_client);
+    if (!r || daemon_res.status != CORE_RPC_STATUS_OK)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
```
Comment (moneromooo-monero) [inline src/wallet/wallet_rpc_server.cpp]: Do other calls do a similar thing with dropping daemon error returns to just give a generic error ?

File: src/wallet/wallet_rpc_server.cpp (hunk 1 above)
Comment (stoffu) [inline src/wallet/wallet_rpc_server.cpp]: This code is copied from [simplewallet.cpp L1608](https://github.com/monero-project/monero/blob/82411cdf3a35cef0518a6431bed016977f5deb50/src/simplewallet/simplewallet.cpp#L1608). Here, 2 is the minimum of threads count and the maximum is the result of `tools::get_max_concurrency()`.

File: src/wallet/wallet_rpc_server.cpp (hunk 2 above)
Comment (stoffu) [inline src/wallet/wallet_rpc_server.cpp]: `WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR` is used pretty much everywhere in `wallet_rpc_server.cpp` where a call to a `wallet2` member function can throw. The slight difference with this code is that it directly calls a daemon RPC.

**Comment by stoffu** [comment]
Yes, it's just a proxy to the daemon RPC. I thought of doing this for the sake of consistency with the wallet cli which has `start_mining` and `stop_mining` commands.

**Comment by stoffu** [comment]
> Or is this more convenience?

This is just for convenience, and I agree that it's not so much meaningful, so I'm happy to withdraw it.

But I guess the same logic also applies to `monero-wallet-cli`: if the wallet CLI is connected to a local daemon, you can just use `start_mining` in `monerod` instead of using `start_mining` in `monero-wallet-cli`.

**Pull request merged**